hamilton/function_modifiers/macros.py

```diff
@@ -876,3 +876,134 @@ def optional_config(self) -> Dict[str, Any]:
+        :param collapse: Whether to collapse this into a single node. This is not currently supported.
+        :param _chain: Whether to chain the first parameter. This is the only mode that is supported. Furthermore, this is not externally exposed. @flow will make use of this.
+        """
+        self.transforms = transforms
+        self.collapse = collapse
+        self.chain = _chain
+        self.namespace = namespace
+
+        if self.collapse:
+            raise NotImplementedError(
+                "Collapsing step() functions as one node is not yet implemented for pipe(). Please reach out if you want this feature."
+            )
+
+        if self.chain:
+            raise NotImplementedError("@flow() is not yet supported -- this is ")
+
+    def transform_dag(
+        self, nodes: Collection[node.Node], config: Dict[str, Any], fn: Callable
+    ) -> Collection[node.Node]:
+        """Transforms the subDAG by getting the injectable parameters (anything not
+        produced by nodes inside it), then calling the inject_nodes function on it.
+
+        :param nodes:
+        :param config:
+        :param fn:
+        :return:
+        """
+
+        # We want the output of post-pipe to connect downstream, so the output needs to be named the
+        # same as the original function. We rename the first function to fn_raw and bind it as the
+        # first parameter. When the pipe nodes are wired through we rename the last node as the
+        # original function name
+        original_node = nodes[-1].copy_with(name=f"{nodes[-1].name}_raw")
+        nodes_to_inject = self.inject_nodes({original_node.name: original_node.type}, config, fn)
+        last_node = nodes_to_inject[-1].copy_with(name=f"{nodes[-1].name}")
+
+        # This is a problem since we loose the typehinting of the last node
+        print(last_node.type)
+        out = nodes[:-1]
+        out.append(original_node)
+        out.extend(nodes_to_inject[:-1])
+        out.append(last_node)
+        return out
+
+    def inject_nodes(
+        self, params: Dict[str, Type[Type]], config: Dict[str, Any], fn: Callable
+    ) -> List[node.Node]:
+        """Injects nodes into the graph. This creates a node for each pipe() step,
+        then reassigns the inputs to pass it in."""
+
+        # I think the better approach would be to draw from the last node another `identity` node
+        # that has the correct name but just passes the value of the last pipe node through. This
+        # would allow us to display all pipe transforms in the DAG, otherwise the last one gets
+        # re-named and we effectively hide a node, but I am stuck at how to dynamically add the
+        # correct typehints
+        def __identity(foo: Any) -> Any:
```

Comment:
So re: typing, we have `input_types` in node (as we discussed earlier) which should help. The `copy_with` allows you to overwrite it.

That said, I'm going back and forth on whether we want an identity node, or we just want to name it with the name of the last one. 
- With `__identity` we preserve the transformations
- With just the name of the node, we simplify the DAG

Thoughts?

Reply:
I am leaning towards keeping the `__identity` to have full transparency in the DAG. I guess I am more comfortable to err on the side of having things a bit cluttered compared to hiding a full transformation of the node.

Looking a bit into the future -- if we implement `collapse`, the user will have the option to simplify the DAG if he wants to.

Let me know if I am missing something about clustering that would mess up UX?

Reply:
As long as it's namespaced, clustering will work. And yep, see the point -- happy to go with it for now!